src/plots/cartesian/axis_defaults.js

```diff
@@ -142,6 +138,11 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
         delete containerOut.zerolinewidth;
     }
 
+    // fill in categories
+    containerOut._initialCategories = axType === 'category' ?
+        orderedCategories(letter, containerOut.categoryorder, containerOut.categoryarray, options.data) :
+        [];
+
     return containerOut;
```

Comment:
As a rule, _user_ data (i.e. `gd.data` / `gd.layout` or their nested `containerIn` objects) should **not** be used anywhere after the supply-defautls step. They are a few exceptions to that rule (namely `scene.camera`), but we should try to keep the number of exceptions as small as possible.

Previously, `orderedCategories` was using `containerIn.categoryorder`. In the case where `categoryorder` was coerced to `'array'` via `catgoryarray` [here](https://github.com/plotly/plotly.js/pull/510/files#diff-3661cf63fdd40306a0d925c49b6c05fdL24), `orderedCategories` was not given the correct `categoryorder` value.

In order words, user data/layout is (exactly) what the user inputs, whereas _full_ data/layout is what plotly.js uses to compute and draw things.